### PR TITLE
Add an algorithm for setting the agent certificate serial number

### PR DESCRIPTION
**Pull request by mfoltzgoogle**
This addresses Issue #276: Agent Certificate has a circular dependency on itself

We can't use the certificate fingerprint as the serial number, because the serial number is embedded in the certificate itself.

X.509 requires every certificate from the same issuer to have a unique (not necessarily consecutive) serial number.

This PR adds a simple algorithm to generate a unique per-agent serial number per certificate.   Even if an agent generates a new certificate once per hour, the counter would take almost 500,000 years to overflow :-P

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/openscreenprotocol/pull/293.html" title="Last updated on Jan 22, 2024, 7:45 PM UTC (a90f8df)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/openscreenprotocol/293/15ed7a5...a90f8df.html" title="Last updated on Jan 22, 2024, 7:45 PM UTC (a90f8df)">Diff</a>

### CODE REVIEW

**Comment by mfoltzgoogle** [comment]
Feel free to review now in advance of TPAC.

**Comment by mfoltzgoogle** [comment]
baylesj@, do you mind taking a look?

**Comment by backkem** [comment]
Is the same comment not relevant to the `Subject Name`? It is defined as `CN = <fp>._openscreen._udp`.

File: index.bs [hunk 1]
```diff
@@ -375,6 +375,26 @@ The [=agent certificate=] must have the following characteristics:
...
+Let the <dfn>certificate serial number</dfn> be the result of the following steps:
+
+<ol>
+  <li>If the agent has never generated an agent certificate:
+  <ol>
+    <li>Let the <dfn>certificate serial number base</dfn> be a 32-bit
+        pseudorandom integer value.</il>
+    <li>Let the <dfn>certificate serial number counter</dfn> be a 32-bit
+        unsigned integer.</li>
+    <li>Assign the [=certificate serial number counter=] to 0.</li>
```
Comment (baylesj) [inline index.bs:387]: Nit: the base is declared and assigned in a single step. For consistency, have both the base and counter either declared and assigned in one step each, or two steps.

**Review by baylesj — APPROVED** [review]
LGTM % feedback.

**Comment by backkem** [comment]
The use of <fp> in the subject name is no issue? Is there a way to set this field after calculation of the fingerprint?

File: index.bs (hunk 1 above)
Comment (mfoltzgoogle) [reply index.bs:389]: Done

**Comment by mfoltzgoogle** [comment]
> Is the same comment not relevant to the `Subject Name`? It is defined as `CN = <fp>._openscreen._udp`.

Yes, that is also an issue, thank you for pointing that out.  We won't be able to close Issue #276 until we address that.

**Comment by mfoltzgoogle** [comment]
PTAL @baylesj, I will have to come up with a separate proposal to fix the subject name.

**Comment by mfoltzgoogle** [comment]
Went ahead and merged this as comments have been addressed. Please let me know if there are other comments @baylesj

**Pull request merged**